src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -0,0 +1,3302 @@
+									  NodeConnectionTimeout)));
+		}
+		else
+		{
+			/* stop interrupting WaitEventSetWait for timeouts */
+			workerPool->checkForPoolTimeout = false;
+		}
+	}
+}
+
+
+/*
+ * NextEventTimeout finds the earliest time at which we need to interrupt
+ * WaitEventSetWait because of a timeout and returns the number of milliseconds
+ * until that event with a minimum of 1ms and a maximum of 1000ms.
+ *
+ * This code may be sensitive to clock jumps, but only has the effect of waking
+ * up WaitEventSetWait slightly earlier to later.
+ */
+static long
+NextEventTimeout(DistributedExecution *execution)
+{
+	ListCell *workerCell = NULL;
+	TimestampTz now = GetCurrentTimestamp();
+	long eventTimeout = 1000; /* milliseconds */
+
+	foreach(workerCell, execution->workerList)
+	{
+		WorkerPool *workerPool = (WorkerPool *) lfirst(workerCell);
+
+		if (workerPool->failed)
+		{
+			/* worker pool may have already timed out */
+			continue;
+		}
+
+		if (workerPool->poolStartTime != 0 && workerPool->checkForPoolTimeout)
+		{
+			long timeSincePoolStartMs =
+				MillisecondsBetweenTimestamps(workerPool->poolStartTime, now);
+
+			/*
+			 * This could go into the negative if the connection timeout just passed.
+			 * In that case we want to wake up as soon as possible. Once the timeout
+			 * has been processed, checkForPoolTimeout will be false so we will skip
+			 * this check.
+			 */
+			long timeUntilConnectionTimeoutMs =
+				NodeConnectionTimeout - timeSincePoolStartMs;
+
+			if (timeUntilConnectionTimeoutMs < eventTimeout)
+			{
+				eventTimeout = timeUntilConnectionTimeoutMs;
+			}
+		}
+	}
+
+	return Max(1, eventTimeout);
+}
+
+
+/*
+ * MillisecondsBetweenTimestamps is a helper to get the number of milliseconds
+ * between timestamps when it is expected to be small enough to fit in a
+ * long.
+ */
+static long
+MillisecondsBetweenTimestamps(TimestampTz startTime, TimestampTz endTime)
+{
+	long secs = 0;
+	int micros = 0;
+
+	TimestampDifference(startTime, endTime, &secs, &micros);
+
+	return secs * 1000 + micros / 1000;
+}
+
+
+/*
+ * ConnectionStateMachine opens a connection and descends into the transaction
+ * state machine when ready.
+ */
+static void
+ConnectionStateMachine(WorkerSession *session)
+{
+	WorkerPool *workerPool = session->workerPool;
+	DistributedExecution *execution = workerPool->distributedExecution;
+
+	MultiConnection *connection = session->connection;
+	MultiConnectionState currentState;
+
+	do {
+		currentState = connection->connectionState;
+
+		switch (currentState)
+		{
+			case MULTI_CONNECTION_INITIAL:
+			{
+				/* simply iterate the state machine */
+				connection->connectionState = MULTI_CONNECTION_CONNECTING;
+				break;
+			}
+
+			case MULTI_CONNECTION_CONNECTING:
+			{
+				PostgresPollingStatusType pollMode;
+
+				ConnStatusType status = PQstatus(connection->pgConn);
+				if (status == CONNECTION_OK)
+				{
+					elog(DEBUG4, "established connection to %s:%d", connection->hostname,
+						 connection->port);
```

Comment:
it would be useful to add the session ID here so we can relate it to other debug messages.